File: Sep24/Lecture9Examples/Example33/WeatherViewModel.java
package com.varunshankar.example33;

import android.app.Application;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;

import androidx.core.os.HandlerCompat;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import org.json.JSONException;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.net.URL;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class WeatherViewModel extends AndroidViewModel {
    private MutableLiveData<WeatherData> jsonData =
            new MutableLiveData<WeatherData>();
    private String mLocation;

    public WeatherViewModel(Application application){
        super(application);
    }

    public void setLocation(String location){
        mLocation = location;
        loadData();
    }

    public LiveData<WeatherData> getData(){
        return jsonData;
    }

     private void loadData(){
        new LoadTask().execute(mLocation);
    }
    private class LoadTask{
        ExecutorService executorService = Executors.newSingleThreadExecutor();
        Handler mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper());
        public void execute(String location)
        {
            executorService.execute(new Runnable(){
                @Override
                public void run(){
                    String jsonWeatherData;
                    URL weatherDataURL = NetworkUtils.buildURLFromString(location);
                    jsonWeatherData = null;
                    try{
                        jsonWeatherData = NetworkUtils.getDataFromURL(weatherDataURL);
                        if(jsonWeatherData!=null)
                            postToMainThread(jsonWeatherData);
                    }catch(Exception e){
                        e.printStackTrace();
                    }
                }
            });
        }

        private void postToMainThread(String retrievedJsonData)
        {
            mainThreadHandler.post(new Runnable() {
                @Override
                public void run() {
                    try {
                        jsonData.setValue(JSONWeatherUtils.getWeatherData(retrievedJsonData));
                    }catch(JSONException e){
                        e.printStackTrace();
                    }
                }
            });
        }
    }

 /*   private static class LoadTask extends AsyncTask<String,Void,String>{
        private WeakReference<WeatherViewModel> mWeatherViewModelWeakReference;

        LoadTask(WeatherViewModel weatherViewModel)
        {
            mWeatherViewModelWeakReference = new WeakReference<WeatherViewModel>(weatherViewModel);
        }

        @Override
        protected String doInBackground(String... strings) {
            String location = strings[0];
            URL weatherDataURL = null;
            String retrievedJsonData = null;
            if(location!=null) {
                weatherDataURL = NetworkUtils.buildURLFromString(location);
                try {
                    retrievedJsonData = NetworkUtils.getDataFromURL(weatherDataURL);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return retrievedJsonData;
        }

        @Override
        protected void onPostExecute(String retrievedJsonData) {
            try {
                if (mWeatherViewModelWeakReference==null) return;
                WeatherViewModel local = mWeatherViewModelWeakReference.get();
                local.jsonData.setValue(JSONWeatherUtils.getWeatherData(retrievedJsonData));
            }catch(JSONException e){
                e.printStackTrace();
            }

        }
    };*/
}

File: Sep24/Example33/app/src/main/java/com/varunshankar/example33/WeatherViewModel.kt
package com.varunshankar.example33

import com.varunshankar.example33.NetworkUtils.buildURLFromString
import com.varunshankar.example33.NetworkUtils.getDataFromURL
import com.varunshankar.example33.JSONWeatherUtils.getWeatherData
import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.LiveData
import androidx.core.os.HandlerCompat
import android.os.Looper
import org.json.JSONException
import java.lang.Exception
import java.util.concurrent.Executors

class WeatherViewModel(application: Application?) : AndroidViewModel(application!!) {
    private val jsonData = MutableLiveData<WeatherData>()
    private var mLocation: String? = null
    fun setLocation(location: String?) {
        mLocation = location
        loadData()
    }

    val data: LiveData<WeatherData>
        get() = jsonData

    private fun loadData() {
        if (mLocation!=null)
            LoadTask().execute(mLocation)
    }

    private inner class LoadTask {
        val executorService = Executors.newSingleThreadExecutor()!!
        val mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper())
        fun execute(location: String?) {
            executorService.execute {
                var jsonWeatherData: String?
                val weatherDataURL = buildURLFromString(location!!)
                jsonWeatherData = null
                try {
                    jsonWeatherData = getDataFromURL(weatherDataURL!!)
                    jsonWeatherData?.let { postToMainThread(it) }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        private fun postToMainThread(retrievedJsonData: String) {
            mainThreadHandler.post {
                try {
                    jsonData.setValue(getWeatherData(retrievedJsonData))
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        }
    } /*   private static class LoadTask extends AsyncTask<String,Void,String>{
        private WeakReference<WeatherViewModel> mWeatherViewModelWeakReference;

        LoadTask(WeatherViewModel weatherViewModel)
        {
            mWeatherViewModelWeakReference = new WeakReference<WeatherViewModel>(weatherViewModel);
        }

        @Override
        protected String doInBackground(String... strings) {
            String location = strings[0];
            URL weatherDataURL = null;
            String retrievedJsonData = null;
            if(location!=null) {
                weatherDataURL = NetworkUtils.buildURLFromString(location);
                try {
                    retrievedJsonData = NetworkUtils.getDataFromURL(weatherDataURL);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return retrievedJsonData;
        }

        @Override
        protected void onPostExecute(String retrievedJsonData) {
            try {
                if (mWeatherViewModelWeakReference==null) return;
                WeatherViewModel local = mWeatherViewModelWeakReference.get();
                local.jsonData.setValue(JSONWeatherUtils.getWeatherData(retrievedJsonData));
            }catch(JSONException e){
                e.printStackTrace();
            }

        }
    };*/
}